lib/Driver/Job.cpp

```diff
@@ -52,6 +73,28 @@ void CommandOutput::ensureEntry(StringRef PrimaryInputFile,
   }
 }
 
+void CommandOutput::checkInvariants() const {
+  types::forAllTypes([&](types::ID Type) {
+      size_t numOutputsOfType = 0;
+      for (auto const &I : Inputs) {
+        assert(!I.Primary.empty());
+        auto const *M = DerivedOutputMap.getOutputMapForInput(I.Primary);
+        if (!M)
+          continue;
+        auto const Out = M->find(Type);
+        if (Out == M->end())
+          continue;
+        assert(!Out->second.empty());
+        ++numOutputsOfType;
+      }
+      assert(Type == PrimaryOutputType ||
```

Comment:
This seems separate. The primary output is also going to be 0, 1, or N, right? It's just N sometimes where it should be 0.

Reply:
Yeah, good catch, this was an artifact of the invariant-check being called at the wrong point vis-a-vis the addition of an element to the map, and I misinterpreted the debug output as suggesting it was a valid pre-or-post state on the external interface. It only happens as an intermediate state, fixed.